### ISSUE TITLE
Authentication credentials stored with wrong domain name

### ISSUE THREAD

**Issue by biomedia-thomas** [symfony/flex#345]
When using custom (satis) repositories with authentication, the flex preloader sometimes asks to store the credentials for github rather than the right repository domain. This was fixed but there seems to be a regression in later versions. Relevant earlier issue is #273, I created a new issue as requested.

Given the following composer.json:
```json
{
    "repositories": [
        {
            "type": "composer",
            "url": "https://custom.repos.example"
        }
    ],
    "require": {
        "custom/package": "^1.0",
        "symfony/flex": "1.0.78",
        "symfony/process": "4.0.7"
    }
}
```
(note the symfony/process addition, this is important to reproduce the bug)
Then, in an **empty** directory with just the above json file:
- Empty composer cache (I use windows and remove the Local and Roaming AppData composer dirs)
- composer install
- Composer will ask for credentials, enter these. It will ask to store credentials for custom.repos.example, enter No. Composer.lock and symfony.lock are generated.
- Remove the vendor dir, clear composer cache again but leave composer.lock and symfony.lock
- composer install
- Enter authentication again
- Result: "Do you want to store credentials for **github.com** in C:/Users/Thomas/AppData/Roaming/Composer/auth.json"

Even though the authentication is asked for custom.repos.example (and authentication does work), it will try to store it for github.com. When Yes is answered, the actual auth.json created is:
```!json
{
    "http-basic": {
        "github.com": {
            "username": null,
            "password": null
        }
    }
}
```

This problem appears to have started with version 1.0.62. Version 1.0.59 works. Version 1.0.60 and 61 have a different problem ("You must be using the interactive console to authenticate"). The latest version 1.0.78 doesn't work either.

**Comment by stof** [comment]
The timeframe from 1.0.59 to 1.0.62 contains the rewrite of the parallel downloader to use curl. so it is most likely an issue in this rewrite.

**Comment by nicolas-grekas** [comment]
@biomedia-thomas does this issue happen if you disable ext-curl?

**Comment by biomedia-thomas** [comment]
Yes, it seems to be independent of the curl extension.

**Comment by nicolas-grekas** [comment]
@biomedia-thomas would you be able to work on this one? I don't have a satis at hand...

**Comment by biomedia-thomas** [comment]
I think I have figured out the issue but don't know how to fix it exactly.

The `ParallelDownloader` seems to run its parent's `RemoteFileSystem::get` multiple times asynchronously, but on the same instance of RemoteFileSystem. However RemoteFileSystem stores a lot of state in its class members, including `$this->originUrl` and `$this->fileUrl`. Although a lot of the code uses `$originUrl` and `$fileUrl` as function parameters, these member variables are used as well.

When flex runs get() for my private repository, then get() right after for a package on packagist, it overrides `$this->originUrl` and `$this->fileUrl` with github. When the private repository fails on 401 (not authenticated), it uses `promptAuthAndRetry` to prompt for a password, but now this wrongly uses the packagist originUrl instead of my own repository.

Other (and future) variable in RemoteRepository may have similar issues (storeAuth, lastHeaders, retryAuthFailure etc.).

As a side node, I also noticed when the CurlDownloader throws a 401 TransportException, it `ParallelDownloader::getRemoteContents` always sees this as an error and unnecessarily falls back to `parent::getRemoteContents` to try again, which of course results in another 401 which then gets handles properly.
